Simulator Workload

sentientwaffle opened the pull request:
- The conductor orchestrates a set of clients.
- The Workload generates requests and validates replies.
- The Auditor stores the expected state of replica's state machines. It uses that state to validate replies from the cluster.
- `committed_requests_max` is set to ~3 WAL wraps, so checkpoint is tested.
- Because state transfer & block recovery aren't implemented, the cluster will not checkpoint until all replicas are ready.

For integration:

- Instead of using a custom state machine to validate state, use the existing message hash chain.
- `Cluster` no longer manages the clients or the `StateChecker`.
- Add `MessagePool.init_capacity`: The `Conductor` needs a `MessagePool` that is neither a client nor a replica.
- Add a `grid_size_max` option to the test `Storage`. Currently this is hard-coded, but eventually it should computed as the upper-bound amount of storage required for the test.
- (The old test `StateMachine` is now dead code, but I'm leaving it in case we ever want to run the VOPR on just consensus, not storage).
- Import `PriorityQueue` from Zig master's std to backport a couple fixes. This can be removed once TB upgrades to Zig 0.10.
- `command=reply` now sets the `timestamp` to match the corresponding `command=prepare` (formerly it was 0). This enables the Auditor to check transfer timeouts.
- Add `Client.on_reply_callback`/`Client.on_reply_context`. To check linearizability, the `Conductor` requires access to the original `request` and `reply` messages.

Misc fixes:

- In `benchmark.sh`, don't overwrite `format`'s logs.
- Update comment in `message_pool.zig`.

sentientwaffle committed:
VOPR: StateChecker: validate in-flight & hashchain

Instead of using a custom state machine to validate state, use the existing message hash chain. (This change is a prelude to testing the real StateMachine in the simulator).

Previously, the VOPR used a custom `StateMachine` that chained `state = hash(state, hash(client_id, message_body))`. This wouldn't work after the log has wrapped and a replica crashes/recovers — the state would need to be recovered from replaying every message, instead of starting from the snapshot. (Or the `state` could need to be stored/restored by snapshots).

- Remove `StateChecker`'s request queues:
  - The comment "We must therefore use our own queue rather than the clients' queues." is no longer applicable, because the state checker now uses only the `client` id to identify the request in the client queue (instead of a possibly-colliding hash of the message body).
- validate `register` messages the same as any other message
- remove the dependency on `Cluster`.
- The hashes printed by the state checker during VOPR runs are now the actual message checksums, instead of being random.
  - (Though the number of transitions printed doesn't match the number of requests sent, since it includes `register`s).
- Rename `transitions` → `committed_messages` to clarify that it doesn't recount transitions after crash/recover.
  - (The impetus for renaming was actually to add and distinguish `committed_requests`).
  - The distinction between the two counters is that:
    - `committed_messages` cares about all commits, including `register` messages.
    - `committed_requests` ignores `register`s, since the simulator wants a target number of requests to arrive.

Extra:
- Stricter assertions in Replica and StateMachine.
- In `vsr.Header.calculate_checksum{,_body}`, call `vsr.checksum()`.

sentientwaffle committed:
VOPR: Test Accounting StateMachine with Workload

NOTE: Currently `committed_requests_max` is not set high enough to trigger a checkpoint, because there are still many test failures just from compaction.

- The conductor orchestrates a set of clients.
- The Workload generates requests and validates replies.
  - `TransferBatchQueue` tracks recently committed batches of transfers. This allows `Workload.on_lookup_transfers` to be stricter when verifying the results: it know exactly which transfers must (or can't) exist.
  - Weight the RNG in favor of creating accounts until at least one `create_accounts` has been sent.
  - Checksum Transfers only after all have been constructed (rather than as they are constructed) because linking backtracks.
- The Auditor stores the expected state of replica's state machines. It uses that state to validate replies from the cluster.
  - From the Auditor's point-of-view, the number of "in-flight" requests isn't just the client's request queue. It additionally includes requests whose replies have arrived _but not yet been processed_ (because of a stalled request with a lower op).
- `IdPermutation`
  - Don't use `ChaCha8IETF` for pseudo-UUID generation. Because the block counter is fixed (necessary to allow decryption without storing any extra info) the encrypted output is nearly sequential.
  - The new "algorithm" for the pseudo-UUID just sandwiches the index in the middle of some random data.
- Assert workload/auditor completion

For integration:

- `Cluster` no longer manages the clients or the `StateChecker`.
- Add `MessagePool.init_capacity`: The `Conductor` needs a `MessagePool` that is neither a client nor a replica.
- Add a `grid_size_max` option to the test `Storage`. Currently this is hard-coded, but eventually it should computed as the upper-bound amount of storage required for the test.
- (The old test `StateMachine` is now dead code, but I'm leaving it in case we ever want to run the VOPR on just consensus, not storage).
- Import `PriorityQueue` from Zig master's std to backport a couple fixes. This can be removed once TB upgrades to Zig 0.10.
- `command=reply` now sets the `timestamp` to match the corresponding `command=prepare` (formerly it was 0). This enables the Auditor to check transfer timeouts.
- Add `Client.on_reply_callback`/`Client.on_reply_context`. To check linearizability, the `Conductor` requires access to the original `request` and `reply` messages.
- Fix `StateChecker`. The header for op=0 doesn't always exist after recovery; it may be corrupt.
- Halt commits until everyone checkpoints
  - As a stopgap until state transfer & grid recovery are implemented, don't checkpoint until all replicas are caught up.
  - This allows:
    - the VOPR to keep testing crashes without worrying that a replica will fall too far behind the cluster
    - while also increasing `committed_requests_max` to allow the WAL to wrap.
- Use huge message timeouts so they never trigger — timeouts aren't implemented in the state machine yet.
- SuperBlock/VOPR: Allow deinit mid-operation: The VOPR calls `deinit()` after simulating a crash. At the time of the "crash", the superblock may have work queued anyway.
- Also, update crash/recovery the workaround in `cluster.zig` to avoid panicking. Unfortunately the workaround doesn't play well with WAL wrapping.

Misc fixes:

- In `benchmark.sh`, don't overwrite `format`'s logs.
- Update comment in `message_pool.zig`.

jorangreef commented on scripts/benchmark.sh:
```diff
@@ -50,7 +50,7 @@ for I in $REPLICAS
     echo "Starting replica $I..."
     FILE="./0_${I}.tigerbeetle.benchmark"
-    ./tigerbeetle start --addresses=3001 "$FILE" > benchmark.log 2>&1 &
+    ./tigerbeetle start --addresses=3001 "$FILE" >> benchmark.log 2>&1 &
```
Good catch! Otherwise we only get the log for the last replica right?

sentientwaffle replied:
Yes

jorangreef commented on src/message_pool.zig:
```diff
@@ -87,11 +87,13 @@ pub const MessagePool = struct {
+    }
 
+    pub fn init_capacity(allocator: mem.Allocator, messages_max: usize) error{OutOfMemory}!MessagePool {
         var ret: MessagePool = .{
```
```suggestion
        var pool: MessagePool = .{
```

Good time to improve this.

jorangreef commented on src/simulator.zig:
```diff
@@ -3,16 +3,23 @@ const builtin = @import("builtin");
 const PartitionMode = @import("test/packet_simulator.zig").PartitionMode;
+const MessageBus = @import("test/message_bus.zig").MessageBus;
+const auditor = @import("test/accounting/auditor.zig");
+const Workload = @import("test/accounting/workload.zig").AccountingWorkloadType(StateMachine);
```
```suggestion
const Workload = @import("test/accounting/workload.zig").WorkloadType(StateMachine);
```

So that different workload generators all expose the same `WorkloadType()` interface.

jorangreef replied:
The rationale is that Simulator shouldn't know about the state machine's domain (i.e. in TigerBeetle's case accounting).

jorangreef commented on src/simulator.zig:
```diff
@@ -30,10 +37,15 @@ pub const log_level: std.log.Level = if (log_state_transitions_only) .info else
 pub const deployment_environment = .simulation;
 
+const accounts_batch_size_max = @divFloor(config.message_size_max - @sizeOf(vsr.Header), @sizeOf(tb.Account));
+const transfers_batch_size_max = @divFloor(config.message_size_max - @sizeOf(vsr.Header), @sizeOf(tb.Transfer));
```
Would be great to have these as `options` that the specific StateMachine exports, to completely decouple the simulator from the domain.

sentientwaffle replied:
Since it needs to be comptime, the `StateMachine` would need access to the `message_body_size_max` as comptime (currently it only has it at runtime). What do you think of passing `StateMachine` a `comptime config: struct { message_body_size_max: usize }`. Then it could export these on its `Options` type as declarations, to avoid having two different things called `options`.

sentientwaffle committed:
VSR: Don't repair already-checkpointed headers

Repairing a header that is already checkpointed isn't necessary, and potentially overwrites a newer header.

`on_repair` uses `repair_header` so it also guards against repairing checkpointed headers.

Also:
- Document op/commit invariants.
- Add more assertions to the `journal.is_empty` case (starting the replica for the first time after `tigerbeetlle format`).
- Document and assert some (temporarily) unreachable states. (Unreachable because until state transfer is implemented, the cluster requires all replicas to commit the checkpoint trigger op before any will checkpoint). The assertions guard against the replica getting too far ahead, which could prevent other replicas from repairing if they crash/recover.

sentientwaffle commented:
Note that this change will require clients to rebuild on the latest `client.zig` to be compatible:

> - `command=reply` now sets the timestamp to match the corresponding `command=prepare` (formerly it was 0). This enables the Auditor to check transfer timeouts.

(Previously `invalid_reply()` expects a zero timestamp, but now it must be nonzero. So incompatible clients will drop messages with a nonzero timestamp).

sentientwaffle committed:
Simulator: Storage: Check read after write

This isn't useful for the SuperBlock/WAL zones because they are formatted.
But it has caught an issue in the Grid that is being fixed in a separate branch.

sentientwaffle committed:
StateMachine: Remove `tick` from interface

Remove a potential determinism footgun (which is currently unused anyway).
This was left over from when `StateMachine` ticked the `Grid`.

Also:
- Remove `exists_with_different_{debits,credits}_{pending,posted}` errors (https://github.com/tigerbeetledb/tigerbeetle/pull/192).
- State machine expects `op ≠ 0`.

jorangreef commented on src/simulator.zig:
```diff
@@ -82,10 +95,16 @@ pub fn main() !void {
 
+    // TODO: When block recovery and state transfer are implemented, remove this flag to allow
+    // crashes to coexist with WAL wraps.
+    const committed_requests_max: usize = config.journal_slot_count * 3;
```
```suggestion
    const requests_committed_max: usize = config.journal_slot_count * 3;
```

Since requests is the MSB, followed by committed and then max.

jorangreef replied:
Your call though, since I see we do `committed_messages` and `committed_requests` below. Although these two names could also be improved? Since registering a session is also a request. Is there a way to reduce these two counts into one?

jorangreef commented on src/test/accounting/auditor.zig:
```diff
@@ -0,0 +1,573 @@
+
+/// Store expected possible results for an in-flight request.
+/// This reply validation takes advantage of the Workload's additional context about the request.
+const InFlight = union(enum) {
```
```suggestion
const Inflight = union(enum) {
```

Since "inflight" is one word.

sentientwaffle replied:
"in-flight" (hyphenated) is also common.

jorangreef commented on src/simulator.zig:
```diff
@@ -435,6 +444,16 @@ fn random_partition_mode(random: std.rand.Random) PartitionMode {
+    return switch (random.uintLessThan(usize, 4)) {
+        0 => .{ .identity = {} },
+        1 => .{ .reflect = {} },
+        2 => .{ .zigzag = {} },
```
Great names. Would be nice to add a description for each.

sentientwaffle replied:
They are documented when they are defined, in `src/test/id.zig`.

jorangreef commented on src/test/conductor.zig:
```diff
@@ -0,0 +1,357 @@
+//! The Conductor coordinates a set of Clients, scheduling requests (generated by the Workload),
+//! and receiving replies (validated by the Workload).
+//!
+//! Replies from the cluster may arrive out-of-order; the Conductor reassembles them in the
+//! correct order (by ascending op number) before passing them into the Workload.
```
This is a clean design.

jorangreef commented on src/test/conductor.zig:
```diff
@@ -0,0 +1,357 @@
+
+            const request_metadata = self.workload.build_request(
+                client_index,
+                @alignCast(@alignOf(vsr.Header), request_message.buffer[@sizeOf(vsr.Header)..config.message_size_max]),
```
Over 100 columns.

jorangreef commented on src/test/conductor.zig:
```diff
@@ -0,0 +1,357 @@
+            };
+        }
+
+        fn on_reply(context: ?*anyopaque, client: *Client, request_message: *Message, reply_message: *Message) void {
```
Over 100 columns—needs a comma! 😁

jorangreef commented on src/test/id.zig:
```diff
@@ -0,0 +1,89 @@
+    identity: void,
+
+    /// Ascending indices become descending ids.
+    reflect: void,
```
```suggestion
    inversion: void,
```

A mirror reflection wouldn't invert the vertical axis, only the horizontal axis?

jorangreef replied:
... Tenet.

jorangreef commented on src/vsr.zig:
```diff
@@ -234,18 +236,16 @@ pub const Header = extern struct {
     pub fn calculate_checksum(self: *const Header) u128 {
         const checksum_size = @sizeOf(@TypeOf(self.checksum));
         assert(checksum_size == 16);
-        var target: [32]u8 = undefined;
```
I think we carry forward a potential alignment bug that's been latent here for some time, also into our new `checksum()` function (good abstraction!).

Since we're casting an array of u8 bytes (align 1) to a u128 (align 16), I think we're just lucky so far?

CC'ing @ifreund.

sentientwaffle replied:
I will change it to

```zig
pub fn checksum(source: []const u8) u128 {
    var target: u128 = undefined;
    std.crypto.hash.Blake3.hash(source, std.mem.asBytes(&target), .{});
    return target;
}
```

ifreund replied:
> Since we're casting an array of u8 bytes (align 1) to a u128 (align 16), I think we're just lucky so far?
> 
> CC'ing @ifreund.

There's no problem here since `@bitCast()` has no alignment requires as it uses value semantics.

jorangreef replied:
Thanks @ifreund

jorangreef commented on src/vsr.zig:
```diff
@@ -234,18 +236,16 @@ pub const Header = extern struct {
     pub fn calculate_checksum_body(self: *const Header, body: []const u8) u128 {
         assert(self.size == @sizeOf(Header) + body.len);
-        const checksum_size = @sizeOf(@TypeOf(self.checksum_body));
-        assert(checksum_size == 16);
```
Let's keep these to mirror what we do in `calculate_checksum()`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -154,14 +160,29 @@ pub fn ReplicaType(
         /// The op number of the latest committed and executed operation (according to the replica):
         /// The replica may have to wait for repairs to complete before commit_min reaches commit_max.
+        ///
+        /// Invariants (not applicable during status=recovering):
+        /// * `replica.commit_min` exists in the Journal.
+        /// * `replica.commit_min ≤ replica.op`
+        /// * `replica.commit_min ≥ replica.op_checkpoint`.
+        /// * never decreases
         commit_min: u64,
 
         /// The op number of the latest committed operation (according to the cluster):
         /// This is the commit number in terms of the VRR paper.
+        ///
+        /// Invariants:
+        /// * `replica.commit_max ≥ replica.commit_min`.
+        /// * never decreases
```
Really like this new `Invariants` list format as a pattern.

sentientwaffle committed:
StateMachine: constants.batch_max.{Operation}

And other vopr-workload review fixes.

sentientwaffle committed:
StateMachine: remove Account.overflows_debits/credits

sentientwaffle committed:
StateMachine: Remove Account.exceeds_debits/credits

sentientwaffle committed:
VOPR: Cleanup

jorangreef commented on src/vsr.zig:
```diff
@@ -957,7 +959,7 @@ pub fn sector_ceil(offset: u64) u64 {
-    return @bitCast(u128, target[0..@sizeOf(u128)].*);
+    var target: u128 = undefined;
+    std.crypto.hash.Blake3.hash(source, std.mem.asBytes(&target), .{});
+    return target;
```
Apologies @sentientwaffle let's rather keep this as is, since it was in fact safe as @ifreund clarified.

I think now with passing a `u128` we're actually passing too small a target, since Blake3 is a 256-bit hash.

sentientwaffle committed:
VSR: Explicitly truncate checksum

jorangreef reviewed (approved):
🚀
